File: client/javascript/src/macro/mpi.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import "reflect-metadata"
import {Ark} from "@/macro/ark";


// 'design:paramtypes'
// 'design:returntype'
export class MPI {

    public readonly name: string;
    public readonly version: string;
    public readonly proto: string;
    public readonly codec: string;
    public readonly flags: number;
    public readonly timeout: number;
    public readonly retries: number;
    public readonly node: string;
    public readonly inst: string;
    public readonly zone: string;
    public readonly cluster: string;
    public readonly cell: string;
    public readonly group: string;
    public readonly address: string;
    public kind: any = undefined;
    public retype: any = undefined;
    public attributes: Map<string, string> = new Map();

    constructor(name: string, version: string, proto: string, codec: string, flags: number, timeout: number,
                retries: number, node: string, inst: string, zone: string, cluster: string, cell: string, group: string,
                address: string, retype: any) {
        this.name = name;
        this.version = version;
        this.proto = proto;
        this.codec = codec;
        this.flags = flags;
        this.timeout = timeout;
        this.retries = retries;
        this.node = node;
        this.inst = inst;
        this.zone = zone;
        this.cluster = cluster;
        this.cell = cell;
        this.group = group;
        this.address = address;
        this.retype = retype;
    }

    decorate<T>(target: Object, propertyKey: string | symbol, descriptor?: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T> | void {
        if (typeof propertyKey === "string") {
            this.attributes.set('name', propertyKey);
        }
        if (!descriptor) {
            this.kind = Reflect.getMetadata('design:type', target, propertyKey);
            Ark.annotate(MPI, propertyKey, target, this);
            return;
        }
        this.kind = Reflect.getMetadata('design:returntype', target, propertyKey)
        Ark.annotate(MPI, descriptor.value, target.constructor, this);
        return descriptor;
    }

}

export function mpi(
    name = '',
    retype: any = Object,
    version = '1.0.0',
    proto = 'grpc',
    codec = 'json',
    flags = 0,
    timeout = 5000,
    retries = 3,
    node = '',
    inst = '',
    zone = '',
    cluster = '',
    cell = '',
    group = '',
    address = '',
): PropertyDecorator & MethodDecorator {
    const meta = new MPI(name, version, proto, codec, flags, timeout, retries, node, inst, zone, cluster, cell,
        group, address, retype);
    return meta.decorate.bind(meta);
}

File: client/golang/macro/mpi.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package macro

import "fmt"

func init() {
	var _ MPI = new(MPIAnnotation)
}

type ServiceInspector interface {
	// Inspect multiple service plex.
	Inspect() []MPI
}

// MPI
// Multi Attr Interface. Mesh provider interface.
type MPI interface {
	// Rtt x
	// Define the reference metadata
	Rtt() *Rtt
}

type MPIAnnotation struct {
	Meta *Rtt
}

func (that *MPIAnnotation) Rtt() *Rtt {
	return that.Meta
}

type Rtt struct {
	// Service name. {@link MPS#name()}
	Name string
	// Service version. {@link MPS#version()}
	Version string
	// Service net/io protocol.
	Proto string
	// Service codec.
	Codec string
	// Service flag 1 asyncable 2 encrypt 4 communal.
	Flags int64
	// Service invoke timeout. millions.
	Timeout int64
	// Invoke retry times.
	Retries int
	// Service node identity.
	Node string
	// Service inst identity.
	Inst string
	// Service zone.
	Zone string
	// Service cluster.
	Cluster string
	// Service cell.
	Cell string
	// Service group.
	Group string
	// Service address.
	Address string
}

func (that *Rtt) String() string {
	return fmt.Sprintf("%s.%s", that.Name, that.Version)
}
